sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/JsonNodeStorePayload.java

```diff
@@ -16,49 +19,107 @@
 import java.nio.charset.CharsetDecoder;
 import java.nio.charset.CodingErrorAction;
 import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.function.Consumer;
 
 public class JsonNodeStorePayload implements StorePayload<JsonNode> {
     private static final Logger logger = LoggerFactory.getLogger(JsonNodeStorePayload.class);
     private static final CharsetDecoder fallbackCharsetDecoder = getFallbackCharsetDecoder();
     private final int responsePayloadSize;
     private final JsonNode jsonValue;
 
-    public JsonNodeStorePayload(ByteBufInputStream bufferStream, int readableBytes) {
+    public JsonNodeStorePayload(
+        ByteBufInputStream bufferStream,
+        int readableBytes,
+        Map<String, String> responseHeaders,
+        Callable<Void> interceptor) throws Exception {
+
         if (readableBytes > 0) {
             this.responsePayloadSize = readableBytes;
-            this.jsonValue = fromJson(bufferStream, readableBytes);
+            this.jsonValue = fromJson(bufferStream, readableBytes, responseHeaders, interceptor);
         } else {
             this.responsePayloadSize = 0;
             this.jsonValue = null;
         }
     }
 
-    private static JsonNode fromJson(ByteBufInputStream bufferStream, int readableBytes) {
+    private static JsonNode fromJson(
+        ByteBufInputStream bufferStream,
+        int readableBytes,
+        Map<String, String> responseHeaders,
+        Callable<Void> interceptor) throws Exception {
+
         byte[] bytes = new byte[readableBytes];
         try {
+
+            if (interceptor != null) {
+                interceptor.call();
+            }
+
             bufferStream.read(bytes);
+
             return Utils.getSimpleObjectMapper().readTree(bytes);
         } catch (IOException e) {
             if (fallbackCharsetDecoder != null) {
                 logger.warn("Unable to parse JSON, fallback to use customized charset decoder.", e);
```

Comment:
We should have this log above if / else, so that we log this IOException as logger.error